File: python/worker.py
import asyncio
import dataclasses
import logging
import os

import temporalio.converter
from temporalio.client import (
    Client,
    ClientConfig,
    TLSConfig,
)
from temporalio.worker import Worker

from account_transfer_workflow import AccountTransferWorkflow
from account_transfer_workflow_scenarios import AccountTransferWorkflowScenarios
from activities import AccountTransferActivities
from codec import EncryptionCodec

async def build_client():
    address = os.getenv("TEMPORAL_ADDRESS", "127.0.0.1:7233")
    namespace = os.getenv("TEMPORAL_NAMESPACE", "default")
    tlsCertPath = os.getenv("TEMPORAL_CERT_PATH", "")
    tlsKeyPath = os.getenv("TEMPORAL_KEY_PATH", "")
    apiKey = os.getenv("TEMPORAL_API_KEY", "")
    encryptPayloads = True if os.getenv("ENCRYPT_PAYLOADS", "").lower() == "true" else False

    config = ClientConfig(
        target_host=address,
        namespace=namespace,
    )

    if encryptPayloads:
        print("Encrypting payloads")
        config["data_converter"] = dataclasses.replace(temporalio.converter.default(), payload_codec=EncryptionCodec())

    # Prefer API Key auth
    if apiKey != "":
        print("Using Cloud API key auth")
        print(f"API Key: {apiKey[:4]}...{apiKey[-4:]}")
        print(f"Address: {address}")
        print(f"Namespace: {namespace}")
        config["api_key"] = apiKey
        config["tls"] = True
    # Fallback to mTLS auth
    elif tlsCertPath and tlsKeyPath:
        print("Using mTLS auth")
        with open(tlsCertPath, "rb") as f:
            cert = f.read()
        with open(tlsKeyPath, "rb") as f:
            key = f.read()
        config["tls"] = TLSConfig(client_cert=cert, client_private_key=key)

    return await Client.connect(**config)

async def main():
    logging.basicConfig(level=logging.INFO,
                        format="%(asctime)s | %(levelname)s | %(filename)s:%(lineno)s | %(message)s")

    taskQueue = os.getenv("TEMPORAL_TASK_QUEUE", "MoneyTransfer")
    address = os.getenv("TEMPORAL_ADDRESS", "127.0.0.1:7233")

    client = await build_client()

    activities = AccountTransferActivities()

    worker = Worker(
        client,
        task_queue=taskQueue,
        workflows=[AccountTransferWorkflow, AccountTransferWorkflowScenarios],
        activities=[activities.validate,
                    activities.withdraw,
                    activities.deposit,
                    activities.sendNotification,
                    activities.undoWithdraw],
    )
    print(f"Connecting to Temporal on {address}")
    print("Python money transfer worker starting...")
    await worker.run()

if __name__ == "__main__":
    asyncio.run(main())

File: ruby/worker.rb
# frozen_string_literal: true

require 'temporalio/client'
require 'temporalio/worker'
require 'logger'
require_relative 'workflows/account_transfer_workflow'
require_relative 'workflows/account_transfer_workflow_scenarios'
require_relative 'activities/validate_activity'
require_relative 'activities/withdraw_activity'
require_relative 'activities/deposit_activity'
require_relative 'activities/send_notification_activity'
require_relative 'activities/undo_withdraw_activity'
require_relative 'security/encryption_codec'

class Worker
  def run
    logger.info('Ruby money transfer worker starting...')

    Temporalio::Worker.new(
      client: temporal_client,
      task_queue: task_queue,
      workflows: [
        Workflows::AccountTransferWorkflow,
        Workflows::AccountTransferWorkflowScenarios
      ],
      activities: [
        Activities::ValidateActivity,
        Activities::WithdrawActivity,
        Activities::DepositActivity,
        Activities::SendNotificationActivity,
        Activities::UndoWithdrawActivity
      ],
      workflow_payload_codec_thread_pool: Temporalio::Worker::ThreadPool.default
    ).run
  end

  private

  def logger
    @logger ||= Logger.new($stdout, level: :info)
  end

  def temporal_client
    @client ||= begin
      options = {
        logger: logger
      }.tap do |options|
        options.merge!(tls_options) if using_tls?
        options.merge!(encryption_options) if encrypt_payloads?
      end
      Temporalio::Client.connect(temporal_address, temporal_namespace, **options)
    end
  end

  def tls_options
    if api_key
      {
        tls: true,
        api_key: api_key,
        rpc_metadata: { 'temporal-namespace' => temporal_namespace }
      }
    elsif cert_path && key_path
      {
        tls: Temporalio::Client::Connection::TLSOptions.new(
          client_cert: File.read(cert_path),
          client_private_key: File.read(key_path)
        )
      }
    else
      {}
    end
  end

  def encryption_options
    return {} unless encrypt_payloads?

    {
      data_converter: Temporalio::Converters::DataConverter.new(
        payload_codec: Security::EncryptionCodec.new
      )
    }
  end

  def temporal_address
    ENV.fetch('TEMPORAL_ADDRESS', 'localhost:7233')
  end

  def temporal_namespace
    ENV.fetch('TEMPORAL_NAMESPACE', 'default')
  end

  def task_queue
    ENV.fetch('TEMPORAL_MONEYTRANSFER_TASKQUEUE', 'MoneyTransfer')
  end

  def api_key
    ENV['TEMPORAL_API_KEY']
  end

  def cert_path
    ENV['TEMPORAL_CERT_PATH']
  end

  def key_path
    ENV['TEMPORAL_KEY_PATH']
  end

  def using_tls?
    api_key || (cert_path && key_path)
  end

  def encrypt_payloads?
    ENV['ENCRYPT_PAYLOADS']&.downcase == 'true'
  end
end

Worker.new.run if __FILE__ == $PROGRAM_NAME
